### PR TITLE
Fix type annotations @ `ansible.galaxy.collection`

### PR DESCRIPTION
**Pull request by webknjaz**
##### SUMMARY

$sbj

##### ISSUE TYPE

- Bugfix Pull Request
- Test Pull Request

##### COMPONENT NAME

lib/ansible/galaxy/collection/__init__.py

##### ADDITIONAL INFORMATION

Refs:
* https://github.com/ansible/ansible/pull/76296#pullrequestreview-806384044
* https://github.com/ansible/ansible/pull/73319

### CODE REVIEW

File: lib/ansible/galaxy/collection/__init__.py
Comment (bcoca) [inline lib/ansible/galaxy/collection/__init__.py:177]: why are we adding an assert here? the norm is to avoid `assert` as it can easily be bypassed by python runtime options

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/galaxy/collection/__init__.py
Comment (webknjaz) [reply lib/ansible/galaxy/collection/__init__.py:177]: Otherwise, mypy "thinks" that it may be `None` and issues an error and it doesn't have `is_concrete_artifact` attribute. It's a common idiom/pattern in projects with the type checks enabled.

This is because the incoming function argument `remote_collection` is typed as `Optional[Candidate]` which means "`None` or `Candidate`". Mypy isn't always able to infer the correct type which calls for a need to use `assert`.

I suppose that a [`cast()`](https://mypy.readthedocs.io/en/stable/casts.html) could be used but it'd require an unconditional `from typing import cast` (py3 stdlib + it's currently not used in runtime + all type annotations still rely on the py2-compatible type comments).

P.S. While ansible-core doesn't yet have this sanity check in the CI, there's a PR I've been working on since refactoring of the dep resolver — it's sorta yet ready and was blocked on the `ansible-test` refactoring but at some point, it'll be possible to enable it and it'd better not produce type errors, at least not for these modules that we've covered with the annotations back then.

Comment (mattclay) [reply lib/ansible/galaxy/collection/__init__.py:177]: An alternative to the `assert` would be to use `if remote_collection is None` instead of `if verify_local_only`.

**Review by mattclay — COMMENTED** [review]
I think I'm leaning towards changing the conditional rather than adding the assert. However, this may be something we should wait on until we've finished the discussions the core team is having on how we want to handle type hints.

**Comment by mattclay** [comment]
Closing, now that `mypy` is tested against ansible-core and tests are passing.

**Pull request closed without merging**
